### ISSUE TITLE
"Review time" statistic lower than actual study time

### ISSUE THREAD

**Issue by gstieglitz** [ankidroid/Anki-Android#4119]
Ever since setting a daily study goal as a new year's project, I've been noticing that the review time bar chart (statistics -> review time) trails behind the actual time studied by ~1/6, every day.

For instance, according to the Ankidroid [front page](http://i.imgur.com/6treIAH.png) and my [daily summary](http://i.imgur.com/Vo2C7Cq.png), I studied for 54 minutes today.
According to my [30-day statistic](http://i.imgur.com/icd303U.png), I studied only for 45 minutes.

The [desktop app also shows the larger number](http://i.imgur.com/mGu1RCl.png).

Does anyone know why this happens? I know it's just a cosmetic issue, but I find the 30-day statistic pretty motivating and it can get frustrating to hit my daily total, only to have the graph tell me that I didn't.

**Comment by PSeitz** [comment]
I noticed the same. A look in the code reveals:

`private final double yAxisStretchFactor = 1.05;`
`PlotSheet plotSheet = new PlotSheet(mFirstElement-0.5, mLastElement + 0.5, 0, mMaxCards*yAxisStretchFactor);`

Looks suspicious

**Comment by jvanprehn** [comment]
I've tried to reproduce this by doing this in Stats.java:calculateDone:

        list = new ArrayList<>();
        list.add(new double[] { 0, 29, 1, 20, 4, 1 });
        list.add(new double[] { -1, 20, 2, 21, 4, 1 });
        list.add(new double[] { -2, 21, 1, 23, 1, 1 });
        list.add(new double[] { -16, 6, 45, 43, 5, 1 });

The result looks just fine:

![anki-review-time-stat](https://cloud.githubusercontent.com/assets/16454627/15269314/263cdeda-19fb-11e6-99d6-751ef20019c8.png)

So I don't think it goes wrong in the plotting part.

If I look at your screenshots I also see a difference between Anki Desktop and Ankidroid on day 9. On Ankidroid there is a peek of height around 77. On Anki Desktop it's below 25.

Do we have the SQLite database to reproduce this on?

**Comment by gstieglitz** [comment]
I agree that it doesn't seem to be a plotting issue. Studying new cards of some decks doesn't result in the same imbalance we see here. I've tried to reproduce this with mixed results.

I'm throwing my database up in case anyone's interested. You could also try to reproduce on your own, since I'm mainly using shared decks (All in one Kanji, Core10k F. Opt.) and at least as reviews are concerned, they all seem equally affected.

Collection: https://www.dropbox.com/s/3asd3gof5abzads/collection.anki2?dl=0

FWIW, I did some date changing to keep my 40min/day streak up, but stopped around the chasm at -10d. Reviews past than that point may look a bit weird.

**Comment by hssm** [comment]
I haven't looked into this in detail yet, but:

The query that builds the data used for this graph is somewhat different from the desktop version and seems to be missing a lot of things from it. (Compare: [Anki](https://github.com/dae/anki/blob/master/anki/stats.py#L468)/[AnkiDroid](https://github.com/ankidroid/Anki-Android/blob/develop/AnkiDroid/src/main/java/com/ichi2/libanki/Stats.java#L536)). It's *probably* the cause for the discrepancy.

At the moment I'm trying to add the remaining values for https://github.com/ankidroid/Anki-Android/issues/3623 and I might end up porting over the original query for it. Maybe after that's done we can rework ChartBuilder to use the results from that and see if it fixes things.

**Comment by jvanprehn** [comment]
With your collection, I can't reproduce it either. For me the graph in Ankidroid looks just the same like in the desktop version:

![review-time-nexus-5x-api-23](https://cloud.githubusercontent.com/assets/16454627/15302164/c95266aa-1bb1-11e6-86fd-3cbd55b67b6e.png)

I used this version with cram on top (https://github.com/jvanprehn/Anki-Android/commits/statistics-legend-order) to make it easier to compare.
